src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -277,35 +394,117 @@ public override void ClearAssetCache()
+                    oldAssets[assetId.Name] = new List<AssetId>();
+                    oldAssets[assetId.Name].Add(assetId);
+                }
+                else if (!oldAssets[assetId.Name].Contains(assetId))
+                {
+                    oldAssets[assetId.Name].Add(assetId);
+                }
+
+                var entry = new AssetCacheEntry();
+                entry.Asset = asset;
+                entry.AssetId = assetId;
+                lookupByAsset[asset] = entry;
+                lookupByAssetId[assetId] = entry;
             }
 
-            CleanUpUnused(oldAssets, unvisitedAssets);
-            assets = oldAssets.Values.OrderBy(x => x.AssetId.FileIdentifier).ThenBy(x => x.AssetId.Guid).ToArray();
+            foreach(var idPair in unvisitedIds)
+            {
+                if (oldAssets.TryGetValue(idPair.Item1, out List<AssetId> ids) &&
+                    ids.Contains(idPair.Item2))
+                {
+                    ids.Remove(idPair.Item2);
+                }
+            }
 
+            lookupByName = oldAssets;
+            var tempAssets = new List<NameEntry>();
+            foreach(var item in lookupByName)
+            {
+                tempAssets.Add(new NameEntry(item.Key, item.Value.OrderBy(a => a.Guid).ThenBy(a => a.FileIdentifier).ToArray()));
+            }
+            assets = tempAssets.OrderBy(a => a.Name).ToArray();
             EditorUtility.SetDirty(this);
 #endif
         }
+
+        public virtual void SaveAssets()
+        {
+#if UNITY_EDITOR
+            EnsureAssetDirectoryExists();
+            if (assets == null ||
+                assets.Length == 0)
+            {
+                Debug.Log("Assets was null or empty, rerunning asset cache update");
+                UpdateAssetCache();
+            }
+
+            if (assets == null ||
+                assets.Length == 0)
+            {
+                Debug.Log("No assets were found.");
+                return;
+            }
+
+            foreach(var nameEntry in assets)
+            {
+                string assetName = GetAssetCachesContentPath($"{GetValidAssetName(nameEntry.Name)}_{this.GetType().Name}", assetContentExtension);
+                AssetCacheContent content = ScriptableObject.CreateInstance<AssetCacheContent>();
+                List<AssetCacheEntry> assetCacheEntries = new List<AssetCacheEntry>();
+                foreach (var assetId in nameEntry.Ids)
+                {
+                    if (!lookupByAssetId.TryGetValue(assetId, out var entry))
+                    {
+                        Debug.LogError($"AssetId did not have a registered Asset: {assetId}");
+                        continue;
+                    }
+
+                    assetCacheEntries.Add(entry);
+                }
+
+                content.AssetCacheEntries = assetCacheEntries.ToArray();
+                Debug.Log($"Creating asset: {content} {assetName}");
+                AssetDatabase.CreateAsset(content, assetName);

```

Comment:
It seems like something needs to delete old assets from the asset cache directory. Otherwise, every time an asset name changes, the old version of the asset will still be in this directory